README.md
# kube-register

Register Kubernetes Kubelet machines with the Kubernetes API server using Fleet data.

## Usage

```
kube-register -h
Usage of ./kube-register:
  -api-endpoint="": kubernetes API endpoint
  -fleet-endpoint="": fleet endpoint
  -healthz-port="10255": the kubelet healthz port
  -metadata="k8s=kubelet": comma-delimited key/value pairs
  -reverse-lookup=false: execute reverse lookup for registering hostnames instead of hosts' public IPs
  -sync-interval=30: sync interval
  -version=false: print version and exit
```

### Example

```
kube-register -metadata="kubelet=true" -fleet-endpoint="http://127.0.0.1:4002" -apiserver-endpoint="http://127.0.0.1:8080"
```

By default kube-register registers new machines with their public IP addresses, found in fleet. By setting `-reverse-lookup=true` kube-register will do a reverse DNS lookup to get the machine's hostname and registers the machine with its hostname instead with its IP address. Make sure the machine's IP resolves to the same hostname that is printed by `hostname -f` on the machine itself.

### Enable the fleet endpoint

The kube-register services requires access to a fleet-endpoint. For example on a CoreOS system, enable the fleet end-point with the following systemd socket file:

```
[Socket]
ListenStream=/var/run/fleet.sock
ListenStream=8000
Service=fleet.service

[Install]
WantedBy=sockets.target
```

## Building

```
mkdir -p "${GOPATH}/src/github.com/kelseyhightower"
cd "${GOPATH}/src/github.com/kelseyhightower"
git clone https://github.com/kelseyhightower/kube-register.git
cd kube-register
godep go build .
```

Slightly smaller binary.

```
CGO_ENABLED=0 GOOS=linux godep go build -a -tags netgo -ldflags '-w' .
```

To build the binary inside a docker container.

```
docker run -v $SRC:/opt/kube-register -i -t google/golang /bin/bash -c "cd /opt/kube-register && go get github.com/tools/godep && godep go build ."
```

Where $SRC is the absolute path to the kube-register directory.


CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.0.4] - 2015-05-11
### Changed
- New `reverse-lookup` flag executes a reverse lookup for registering minions via hostnames instead of the public IP.

## [0.0.3] - 2015-04-21  
### Changed
- The default kubelet healthz port is now set to 10255.
  Use the `--healthz-port` to change it.
- Nodes are now registered using the v1beta3 API.
- Don't log when nodes are already registered.


machines.go
package main

import (
	"errors"
	"fmt"
	"log"
	"net"
	"net/http"
	"net/url"

	"github.com/coreos/fleet/client"
)

func getMachines(endpoint, healthzPort string, metadata map[string][]string, reverseLookup bool) ([]string, error) {
	dialFunc := net.Dial
	machineList := make([]string, 0)
	u, err := url.Parse(endpoint)
	if err != nil {
		return nil, err
	}
	if u.Scheme == "unix" {
		endpoint = "http://domain-sock/"
		dialFunc = func(network, addr string) (net.Conn, error) {
			return net.Dial("unix", u.Path)
		}
	}
	c := &http.Client{
		Transport: &http.Transport{
			Dial:              dialFunc,
			DisableKeepAlives: true,
		},
	}
	fleetClient, err := client.NewHTTPClient(c, endpoint)
	if err != nil {
		return nil, err
	}
	machines, err := fleetClient.Machines()
	if err != nil {
		return nil, err
	}
	for _, m := range machines {
		if hasMetadata(m, metadata) && isHealthy(m.PublicIP, healthzPort) {
			if reverseLookup {
				hostnames, err := net.LookupAddr(m.PublicIP)
				if err != nil {
					return nil, err
				}
				if len(hostnames) < 1 {
					return nil, errors.New("Could not get hostname for IP " + m.PublicIP)
				}
				hostname := hostnames[0]
				// remove trailing dot
				if hostname[len(hostname)-1] == '.' {
					hostname = hostname[:len(hostname)-1]
				}
				machineList = append(machineList, hostname)
			} else {
				machineList = append(machineList, m.PublicIP)
			}
		}
	}
	return machineList, nil
}

func isHealthy(addr, healthzPort string) bool {
	url := fmt.Sprintf("http://%s:%s/healthz", addr, healthzPort)
	res, err := http.Get(url)
	if err != nil {
		log.Printf("error health checking %s: %s", addr, err)
		return false
	}
	defer res.Body.Close()
	if res.StatusCode >= http.StatusOK && res.StatusCode < http.StatusBadRequest {
		return true
	}
	log.Printf("unhealthy machine: %s will not be registered", addr)
	return false
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"
)

var (
	apiEndpoint   string
	fleetEndpoint string
	metadata      string
	syncInterval  int
	healthzPort   string
	reverseLookup bool
	printVersion  bool
)

func init() {
	log.SetFlags(0)
	flag.StringVar(&apiEndpoint, "api-endpoint", "", "kubernetes API endpoint")
	flag.StringVar(&fleetEndpoint, "fleet-endpoint", "", "fleet endpoint")
	flag.StringVar(&metadata, "metadata", "k8s=kubelet", "comma-delimited key/value pairs")
	flag.StringVar(&healthzPort, "healthz-port", "10255", "the kubelet healthz port")
	flag.IntVar(&syncInterval, "sync-interval", 30, "sync interval")
	flag.BoolVar(&printVersion, "version", false, "print version and exit")
	flag.BoolVar(&reverseLookup, "reverse-lookup", false, "execute reverse lookup for registering hostnames instead of hosts' public IPs")
}

func main() {
	flag.Parse()
	if printVersion {
		fmt.Printf("kube-register %s\n", Version)
		os.Exit(0)
	}

	m, err := parseMetadata(metadata)
	if err != nil {
		log.Println(err)
	}
	signalChan := make(chan os.Signal, 1)
	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
	for {
		machines, err := getMachines(fleetEndpoint, healthzPort, m, reverseLookup)
		if err != nil {
			log.Println(err)
		}
		for _, machine := range machines {
			if err := register(apiEndpoint, machine); err != nil {
				log.Println(err)
			}
		}
		select {
		case c := <-signalChan:
			log.Println(fmt.Sprintf("captured %v exiting...", c))
			os.Exit(0)
		case <-time.After(time.Duration(syncInterval) * time.Second):
			// Continue syncing machines.
		}
	}
}
